# Proyecto LLM Industrial
## Conexión Dataset MGG001 - Hugging Face

Objetivo:
Cargar datos de equipos industriales, sensores y eventos de falla para análisis predictivo.

In [15]:
import pandas as pd
from huggingface_hub import hf_hub_download

print("Librerías cargadas correctamente")

Librerías cargadas correctamente


In [16]:
reg_path = hf_hub_download(
    repo_id="xpertsystems/mgg001-sample",
    filename="mgg001_equipment_registry.csv",
    repo_type="dataset",
)

registry = pd.read_csv(reg_path)

registry.head()

,asset_id,plant_id,line_id,cell_id,asset_type,manufacturer,model_number,serial_number,installation_date,asset_age_years,...,last_maintenance_date,maintenance_interval_days,days_since_maintenance,cumulative_operating_hours,nominal_speed_rpm,rated_power_kw,bearing_model,lubricant_type,protocol,gateway_id
0,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,KSB,SCH-2152-84,SN-20200127-50534,2020-01-27,6.21,...,2026-02-06,90,62,46448.0,1583.0,97.7,NU205,oil_mineral_iso_vg46,IO_Link,GW-110
1,ASSET-PUM-0002,PLANT-008,LINE-E,CELL-007,pump_centrifugal,ABB,NID-5010-89,SN-20180726-71012,2018-07-26,7.71,...,2026-04-04,90,5,45887.0,2691.0,317.7,6205,oil_mineral_iso_vg46,Profibus,GW-031
2,ASSET-CNC-0003,PLANT-008,LINE-B,CELL-009,cnc_machining_center,ABB,SUL-5280-39,SN-20160205-30421,2016-02-05,10.18,...,2026-03-21,60,19,81124.0,4760.0,49.6,NU205,oil_mineral_iso_vg100,HART,GW-020
3,ASSET-CON-0004,PLANT-002,LINE-E,CELL-004,conveyor_belt,Bosch Rexroth,SIE-8170-80,SN-20150426-80232,2015-04-26,10.96,...,2026-03-17,60,23,76871.0,122.0,136.8,6206,oil_mineral_iso_vg220,Profibus,GW-114
4,ASSET-PUM-0005,PLANT-006,LINE-D,CELL-011,pump_centrifugal,WEG,ABB-6032-80,SN-20240414-37355,2024-04-14,1.99,...,2026-02-26,90,42,12706.0,2298.0,383.7,6206,grease_polyurea,Modbus_TCP,GW-199


In [17]:
sensor_path = hf_hub_download(
    repo_id="xpertsystems/mgg001-sample",
    filename="mgg001_sensor_data.csv",
    repo_type="dataset",
)


sensor = pd.read_csv(sensor_path)


sensor.head()

,asset_id,plant_id,line_id,cell_id,asset_type,observation_timestamp,shift_id,observation_frequency_hz,temp_bearing_drive_end_c,temp_bearing_non_drive_end_c,...,alarm_level,alert_triggered_flag,alarm_triggered_flag,shutdown_triggered_flag,maintenance_recommendation,sensor_data_quality,protocol,gateway_id,facility_temp_c,facility_humidity_rh_pct
0,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,2026-04-09T15:14:38,Night,0.000278,37.61,32.88,...,normal,False,False,False,none,good,IO_Link,GW-110,27.20,46.5
1,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,2026-04-09T16:14:38,Night,0.000278,46.68,43.07,...,normal,False,False,False,none,good,IO_Link,GW-110,24.16,39.6
2,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,2026-04-09T17:14:38,Night,0.000278,52.50,46.16,...,normal,False,False,False,none,good,IO_Link,GW-110,25.71,44.3
3,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,2026-04-09T18:14:38,Night,0.000278,56.98,53.18,...,normal,False,False,False,none,good,IO_Link,GW-110,26.93,65.6
4,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,2026-04-09T19:14:38,Night,0.000278,59.55,56.79,...,normal,False,False,False,none,good,IO_Link,GW-110,27.64,47.1


In [18]:
failure_path = hf_hub_download(
    repo_id="xpertsystems/mgg001-sample",
    filename="mgg001_failure_events.csv",
    repo_type="dataset"
)

failure_path

'C:\\Users\\Stiven\\.cache\\huggingface\\hub\\datasets--xpertsystems--mgg001-sample\\snapshots\\fa42c6c62c938b4169868657bef2bb578e2e74ea\\mgg001_failure_events.csv'

In [19]:
registry = pd.read_csv(reg_path)

sensor = pd.read_csv(sensor_path)

failures = pd.read_csv(failure_path)

In [20]:
print("Equipos:", registry.shape)
print("Sensores:", sensor.shape)
print("Fallas:", failures.shape)

Equipos: (12, 25)
Sensores: (12960, 79)
Fallas: (1, 7)


In [21]:
registry.head()

,asset_id,plant_id,line_id,cell_id,asset_type,manufacturer,model_number,serial_number,installation_date,asset_age_years,...,last_maintenance_date,maintenance_interval_days,days_since_maintenance,cumulative_operating_hours,nominal_speed_rpm,rated_power_kw,bearing_model,lubricant_type,protocol,gateway_id
0,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,KSB,SCH-2152-84,SN-20200127-50534,2020-01-27,6.21,...,2026-02-06,90,62,46448.0,1583.0,97.7,NU205,oil_mineral_iso_vg46,IO_Link,GW-110
1,ASSET-PUM-0002,PLANT-008,LINE-E,CELL-007,pump_centrifugal,ABB,NID-5010-89,SN-20180726-71012,2018-07-26,7.71,...,2026-04-04,90,5,45887.0,2691.0,317.7,6205,oil_mineral_iso_vg46,Profibus,GW-031
2,ASSET-CNC-0003,PLANT-008,LINE-B,CELL-009,cnc_machining_center,ABB,SUL-5280-39,SN-20160205-30421,2016-02-05,10.18,...,2026-03-21,60,19,81124.0,4760.0,49.6,NU205,oil_mineral_iso_vg100,HART,GW-020
3,ASSET-CON-0004,PLANT-002,LINE-E,CELL-004,conveyor_belt,Bosch Rexroth,SIE-8170-80,SN-20150426-80232,2015-04-26,10.96,...,2026-03-17,60,23,76871.0,122.0,136.8,6206,oil_mineral_iso_vg220,Profibus,GW-114
4,ASSET-PUM-0005,PLANT-006,LINE-D,CELL-011,pump_centrifugal,WEG,ABB-6032-80,SN-20240414-37355,2024-04-14,1.99,...,2026-02-26,90,42,12706.0,2298.0,383.7,6206,grease_polyurea,Modbus_TCP,GW-199


In [22]:
registry.info()

<class 'pandas.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   asset_id                    12 non-null     str    
 1   plant_id                    12 non-null     str    
 2   line_id                     12 non-null     str    
 3   cell_id                     12 non-null     str    
 4   asset_type                  12 non-null     str    
 5   manufacturer                12 non-null     str    
 6   model_number                12 non-null     str    
 7   serial_number               12 non-null     str    
 8   installation_date           12 non-null     str    
 9   asset_age_years             12 non-null     float64
 10  design_life_years           12 non-null     int64  
 11  life_consumed_pct           12 non-null     float64
 12  criticality_class           12 non-null     str    
 13  iso_protection_class        12 non-null     str 

In [23]:
sensor.info()

<class 'pandas.DataFrame'>
RangeIndex: 12960 entries, 0 to 12959
Data columns (total 79 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   asset_id                      12960 non-null  str    
 1   plant_id                      12960 non-null  str    
 2   line_id                       12960 non-null  str    
 3   cell_id                       12960 non-null  str    
 4   asset_type                    12960 non-null  str    
 5   observation_timestamp         12960 non-null  str    
 6   shift_id                      12960 non-null  str    
 7   observation_frequency_hz      12960 non-null  float64
 8   temp_bearing_drive_end_c      12960 non-null  float64
 9   temp_bearing_non_drive_end_c  12960 non-null  float64
 10  temp_motor_winding_c          12960 non-null  float64
 11  temp_coolant_inlet_c          12960 non-null  float64
 12  temp_coolant_outlet_c         12960 non-null  float64
 13  temp_ambient

In [24]:
datasets = {
    "registry": registry,
    "sensor": sensor,
    "failures": failures
}


for name, df in datasets.items():
    print("\nDataset:", name)
    print(df.columns.tolist())


Dataset: registry
['asset_id', 'plant_id', 'line_id', 'cell_id', 'asset_type', 'manufacturer', 'model_number', 'serial_number', 'installation_date', 'asset_age_years', 'design_life_years', 'life_consumed_pct', 'criticality_class', 'iso_protection_class', 'atex_zone', 'last_maintenance_date', 'maintenance_interval_days', 'days_since_maintenance', 'cumulative_operating_hours', 'nominal_speed_rpm', 'rated_power_kw', 'bearing_model', 'lubricant_type', 'protocol', 'gateway_id']

Dataset: sensor
['asset_id', 'plant_id', 'line_id', 'cell_id', 'asset_type', 'observation_timestamp', 'shift_id', 'observation_frequency_hz', 'temp_bearing_drive_end_c', 'temp_bearing_non_drive_end_c', 'temp_motor_winding_c', 'temp_coolant_inlet_c', 'temp_coolant_outlet_c', 'temp_ambient_c', 'temp_delta_bearing_c', 'thermal_gradient_c_per_hr', 'vib_overall_rms_mm_s', 'vib_peak_mm_s', 'vib_crest_factor', 'vib_kurtosis', 'vib_1x_rpm_amplitude', 'vib_2x_rpm_amplitude', 'vib_bpfo_amplitude', 'vib_bpfi_amplitude', 'vib_

In [25]:
from itertools import combinations


for (name1, df1), (name2, df2) in combinations(datasets.items(), 2):

    comunes = set(df1.columns).intersection(set(df2.columns))

    print("\n----------------------------")
    print(name1, "vs", name2)
    print("Columnas comunes:")
    
    if comunes:
        for col in comunes:
            print("-", col)
    else:
        print("No hay coincidencias")


----------------------------
registry vs sensor
Columnas comunes:
- protocol
- cell_id
- asset_id
- line_id
- asset_type
- plant_id
- gateway_id

----------------------------
registry vs failures
Columnas comunes:
- asset_id
- asset_type
- criticality_class

----------------------------
sensor vs failures
Columnas comunes:
- asset_id
- asset_type


# Modelo Relacional del Dataset Industrial MGG001

A partir del análisis de columnas comunes entre los tres datasets (`registry`, `sensor` y `failures`), se identificó `asset_id` como la clave principal de integración entre las tablas.

La estructura relacional encontrada representa un modelo típico de gestión de activos industriales:

```text
                         ┌─────────────────────────┐
                         │        REGISTRY          │
                         │   Maestro de equipos     │
                         ├─────────────────────────┤
                         │ PK asset_id              │
                         │ plant_id                 │
                         │ line_id                  │
                         │ cell_id                  │
                         │ asset_type               │
                         │ manufacturer             │
                         │ model_number             │
                         │ serial_number            │
                         │ asset_age_years          │
                         │ life_consumed_pct        │
                         │ criticality_class        │
                         │ maintenance_data         │
                         │ operating_hours          │
                         └────────────┬────────────┘
                                      │
                                      │ 1 : N
                                      │
                 ┌────────────────────┴────────────────────┐
                 │                                         │
                 ▼                                         ▼

     ┌─────────────────────────┐             ┌─────────────────────────┐
     │         SENSOR          │             │        FAILURES         │
     │  Datos de sensores IoT  │             │ Eventos históricos      │
     ├─────────────────────────┤             ├─────────────────────────┤
     │ FK asset_id             │             │ FK asset_id             │
     │ observation_timestamp   │             │ failure_timestamp        │
     │ temperature             │             │ failure_type             │
     │ vibration               │             │ failure_description      │
     │ pressure                │             │ severity                 │
     │ current                 │             │ failure_code             │
     │ other_measurements      │             │                          │
     └─────────────────────────┘             └─────────────────────────┘



## Relaciones identificadas

### REGISTRY → SENSOR

Relación:
**1 : N**

Un equipo industrial puede generar múltiples registros de sensores a través del tiempo.


Ejemplo:

Un motor `MOTOR_001` puede tener miles de mediciones de:

- temperatura
- vibración
- presión
- corriente


---

### REGISTRY → FAILURES

Relación:
**1 : N**

Un equipo puede presentar múltiples eventos de falla durante su ciclo de vida.


Ejemplo:

Un motor `MOTOR_001` puede tener miles de mediciones de:

- temperatura
- vibración
- presión
- corriente


---

### REGISTRY → FAILURES

Relación:
**1 : N**

Un equipo puede presentar múltiples eventos de falla durante su ciclo de vida.


## sensor.asset_id = failures.asset_id


Las mediciones de sensores permiten encontrar patrones previos a un evento de falla.

Ejemplo:

Antes de una falla:

Temperatura ↑
Vibración ↑
Horas operación ↑


pueden indicar una condición anómala.



In [26]:
sensor.head()

,asset_id,plant_id,line_id,cell_id,asset_type,observation_timestamp,shift_id,observation_frequency_hz,temp_bearing_drive_end_c,temp_bearing_non_drive_end_c,...,alarm_level,alert_triggered_flag,alarm_triggered_flag,shutdown_triggered_flag,maintenance_recommendation,sensor_data_quality,protocol,gateway_id,facility_temp_c,facility_humidity_rh_pct
0,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,2026-04-09T15:14:38,Night,0.000278,37.61,32.88,...,normal,False,False,False,none,good,IO_Link,GW-110,27.20,46.5
1,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,2026-04-09T16:14:38,Night,0.000278,46.68,43.07,...,normal,False,False,False,none,good,IO_Link,GW-110,24.16,39.6
2,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,2026-04-09T17:14:38,Night,0.000278,52.50,46.16,...,normal,False,False,False,none,good,IO_Link,GW-110,25.71,44.3
3,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,2026-04-09T18:14:38,Night,0.000278,56.98,53.18,...,normal,False,False,False,none,good,IO_Link,GW-110,26.93,65.6
4,ASSET-CNC-0001,PLANT-008,LINE-E,CELL-014,cnc_machining_center,2026-04-09T19:14:38,Night,0.000278,59.55,56.79,...,normal,False,False,False,none,good,IO_Link,GW-110,27.64,47.1


# Ingeniería de Variables Térmicas y Preparación de Datos de Sensores

## Objetivo

La etapa de ingeniería de variables busca transformar los datos originales provenientes de sensores industriales en información útil para modelos de detección de anomalías y predicción de fallas.

Los datos de sensores contienen mediciones dinámicas del comportamiento del equipo a través del tiempo, por lo cual es necesario realizar procesos de:

- Conversión de variables temporales.
- Validación de calidad de datos.
- Creación de indicadores térmicos.
- Extracción de características relacionadas con degradación del equipo.

Estas nuevas variables permitirán que los modelos de Machine Learning identifiquen patrones asociados con condiciones anormales de operación.

---

# 1. Conversión de variables temporales

La columna `observation_timestamp` representa el instante en que fue registrada cada medición del equipo.

Esta variable inicialmente se encuentra almacenada como texto (`string`), por lo cual debe convertirse al formato fecha/hora de Python (`datetime`).

La conversión permite realizar análisis como:

- Evolución histórica de temperatura.
- Tendencias de degradación.
- Comparación entre periodos de operación.
- Ordenamiento cronológico de las mediciones.

Ejemplo:

Antes:


2026-04-09T15:14:38

Después:

2026-04-09 16:14:38

---

# 2. Variables térmicas del equipo

Las variables de temperatura son indicadores fundamentales en mantenimiento predictivo, especialmente en equipos rotativos donde los rodamientos representan componentes críticos.

El dataset contiene mediciones de:

### Temperatura del rodamiento lado transmisión



### Conversión formato fecha

In [27]:
# Conversión de timestamp a formato datetime

sensor["observation_timestamp"] = pd.to_datetime(
    sensor["observation_timestamp"]
)

sensor["observation_timestamp"].head()

0   2026-04-09 15:14:38
1   2026-04-09 16:14:38
2   2026-04-09 17:14:38
3   2026-04-09 18:14:38
4   2026-04-09 19:14:38
Name: observation_timestamp, dtype: datetime64[us]

In [28]:
# Diferencia de temperatura entre ambos rodamientos

sensor["bearing_temp_difference"] = (
    sensor["temp_bearing_drive_end_c"]
    -
    sensor["temp_bearing_non_drive_end_c"]
)


sensor[
    [
        "temp_bearing_drive_end_c",
        "temp_bearing_non_drive_end_c",
        "bearing_temp_difference"
    ]
].head()

,temp_bearing_drive_end_c,temp_bearing_non_drive_end_c,bearing_temp_difference
0,37.61,32.88,4.73
1,46.68,43.07,3.61
2,52.50,46.16,6.34
3,56.98,53.18,3.80
4,59.55,56.79,2.76


In [29]:
# Diferencia entre temperatura del rodamiento y temperatura ambiente

sensor["thermal_load"] = (
    sensor["temp_bearing_drive_end_c"]
    -
    sensor["facility_temp_c"]
)


sensor[
    [
        "temp_bearing_drive_end_c",
        "facility_temp_c",
        "thermal_load"
    ]
].head()

,temp_bearing_drive_end_c,facility_temp_c,thermal_load
0,37.61,27.20,10.41
1,46.68,24.16,22.52
2,52.50,25.71,26.79
3,56.98,26.93,30.05
4,59.55,27.64,31.91


In [30]:
sensor.info()

<class 'pandas.DataFrame'>
RangeIndex: 12960 entries, 0 to 12959
Data columns (total 81 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   asset_id                      12960 non-null  str           
 1   plant_id                      12960 non-null  str           
 2   line_id                       12960 non-null  str           
 3   cell_id                       12960 non-null  str           
 4   asset_type                    12960 non-null  str           
 5   observation_timestamp         12960 non-null  datetime64[us]
 6   shift_id                      12960 non-null  str           
 7   observation_frequency_hz      12960 non-null  float64       
 8   temp_bearing_drive_end_c      12960 non-null  float64       
 9   temp_bearing_non_drive_end_c  12960 non-null  float64       
 10  temp_motor_winding_c          12960 non-null  float64       
 11  temp_coolant_inlet_c          12960 non

Feature Selection

la variable objetivo

podríamos clasificar como:

0 = comportamiento normal
1 = anomalía

qué variable será objetivo (target),
qué variables pueden entrar como predictores,
qué información está permitida usar,
qué tipo de algoritmo necesitamos,
cómo vamos a evaluar si funciona.

¿Podemos identificar, a partir de las variables operacionales del equipo, si el activo está entrando en una condición anómala?

Hipótesis H1

Las variables capturadas por los sensores industriales contienen patrones suficientes para diferenciar entre un estado normal y un estado anómalo del equipo.

¿Existe realmente una diferencia entre estados?

In [31]:
sensor["anomaly_label"].value_counts()

anomaly_label
0    10568
1     2392
Name: count, dtype: int64

Tenemos anomalías suficientes.

## ¿La etiqueta representa una falla real?

In [32]:
sensor[
[
"anomaly_label",
"failure_mode_active",
"failure_mode_severity"
]
].head()

,anomaly_label,failure_mode_active,failure_mode_severity
0,0,imbalance,none
1,0,imbalance,none
2,0,imbalance,none
3,0,imbalance,none
4,0,imbalance,none


In [33]:
sensor[
[
"anomaly_label",
"failure_mode_active",
"failure_mode_severity"
]
].value_counts()

anomaly_label  failure_mode_active      failure_mode_severity
0              none                     none                     7560
               electrical_fault_stator  none                      569
               cavitation               none                      502
                                        incipient                 393
               imbalance                none                      389
1              gear_tooth_crack         moderate                  342
                                        minor                     270
0              mechanical_looseness     none                      232
               gear_tooth_crack         incipient                 216
1              imbalance                moderate                  203
               mechanical_looseness     moderate                  201
               electrical_fault_stator  minor                     187
0              gear_tooth_crack         none                      176
1              cavitation   

## Planteamiento del problema y comprensión de la condición de falla

Los equipos industriales no suelen presentar una falla súbita; en la mayoría de los casos, el deterioro aparece progresivamente mediante cambios medibles en sus variables operacionales. El análisis inicial del dataset revela que la condición del activo no puede interpretarse únicamente como un estado binario de funcionamiento o falla, ya que existen diferentes etapas de degradación antes de alcanzar una condición crítica. Variables como temperatura de rodamientos, vibración, comportamiento eléctrico, condiciones de lubricación, carga operacional y parámetros del proceso permiten observar la evolución del equipo desde una operación estable hasta una condición anómala que puede comprometer su disponibilidad.

El análisis de las etiquetas disponibles evidencia que `failure_mode_active` identifica el mecanismo de degradación presente en el equipo, mientras que `failure_mode_severity` representa el nivel de afectación alcanzado. Por su parte, `anomaly_label` permite diferenciar aquellos estados donde la condición operacional supera los límites normales y requiere atención. Esto indica que una falla industrial no debe entenderse únicamente como un evento final, sino como una transición progresiva donde aparecen señales tempranas asociadas a desgaste mecánico, desbalance, problemas eléctricos, cavitación u otras condiciones anormales.

Por lo tanto, el objetivo del modelo predictivo no será simplemente identificar si un equipo está fallado, sino aprender los patrones de comportamiento que anteceden una condición anómala, permitiendo generar alertas tempranas y apoyar decisiones de mantenimiento preventivo. La hipótesis inicial plantea que las variables capturadas por los sensores industriales contienen suficiente información para identificar cambios en el estado de salud del activo antes de que ocurra una falla crítica, reduciendo el riesgo de paradas no planificadas y mejorando la gestión del ciclo de vida del equipo.

# Propuesta de solución: Modelo de detección temprana de condiciones anómalas

A partir del análisis realizado, se plantea desarrollar un modelo de aprendizaje automático supervisado cuyo objetivo sea identificar condiciones anómalas en equipos industriales antes de que evolucionen hacia una falla crítica.

La solución propuesta utiliza las variables provenientes de los sensores industriales como fuente principal de información, integrando parámetros térmicos, vibracionales, eléctricos, hidráulicos y de lubricación. Estas variables representan el comportamiento físico del activo durante su operación y permiten capturar señales tempranas asociadas con procesos de degradación.

El modelo tendrá como variable objetivo:


- `0` representa una condición operacional controlada.
- `1` representa una condición anómala que requiere atención.

Definir variable objetivo

In [34]:
# Variable objetivo del modelo

target = "anomaly_label"

sensor[target].value_counts()

anomaly_label
0    10568
1     2392
Name: count, dtype: int64

Definir variables candidatas

In [35]:
features = [

    # Variables térmicas
    "temp_bearing_drive_end_c",
    "temp_bearing_non_drive_end_c",
    "temp_motor_winding_c",
    "thermal_gradient_c_per_hr",
    "temp_delta_bearing_c",
    "thermal_load",

    # Vibración
    "vib_overall_rms_mm_s",
    "vib_peak_mm_s",
    "vib_crest_factor",
    "vib_kurtosis",
    "vib_bpfo_amplitude",
    "vib_bpfi_amplitude",

    # Variables hidráulicas
    "pressure_differential_bar",
    "pressure_lube_oil_bar",
    "flow_rate_m3_hr",
    "cavitation_index",

    # Variables eléctricas
    "current_imbalance_pct",
    "power_factor",
    "active_power_kw",
    "motor_load_pct",

    # Lubricación
    "oil_viscosity_cst",
    "oil_contamination_ntu",
    "hours_since_lubrication",

    # Operación
    "operating_speed_rpm",
    "efficiency_pct"
]

In [36]:
model_data = sensor[
    features + [target]
].copy()


model_data.head()

,temp_bearing_drive_end_c,temp_bearing_non_drive_end_c,temp_motor_winding_c,thermal_gradient_c_per_hr,temp_delta_bearing_c,thermal_load,vib_overall_rms_mm_s,vib_peak_mm_s,vib_crest_factor,vib_kurtosis,...,current_imbalance_pct,power_factor,active_power_kw,motor_load_pct,oil_viscosity_cst,oil_contamination_ntu,hours_since_lubrication,operating_speed_rpm,efficiency_pct,anomaly_label
0,37.61,32.88,62.43,0.082,12.23,10.41,1.3299,3.4471,2.592,2.679,...,2.703,0.8358,46.798,47.90,45.32,9.28,902.8,758.6,84.69,0
1,46.68,43.07,78.30,9.151,21.13,22.52,1.1535,2.8962,2.511,2.609,...,1.519,0.8790,46.186,47.27,56.71,3.10,932.2,745.8,84.69,0
2,52.50,46.16,78.67,5.901,26.45,26.79,0.7931,2.0955,2.642,2.749,...,4.248,0.9473,30.621,31.34,50.83,3.85,2315.1,468.7,84.69,0
3,56.98,53.18,84.07,4.195,30.14,30.05,1.0421,3.1083,2.983,3.033,...,2.257,0.8379,38.166,39.06,44.86,6.16,1795.1,586.5,84.69,0
4,59.55,56.79,87.12,2.535,31.68,31.91,1.2432,3.6768,2.958,2.771,...,4.361,0.8689,47.114,48.22,54.46,0.12,1791.6,727.6,84.69,0


In [37]:
model_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 12960 entries, 0 to 12959
Data columns (total 26 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   temp_bearing_drive_end_c      12960 non-null  float64
 1   temp_bearing_non_drive_end_c  12960 non-null  float64
 2   temp_motor_winding_c          12960 non-null  float64
 3   thermal_gradient_c_per_hr     12960 non-null  float64
 4   temp_delta_bearing_c          12960 non-null  float64
 5   thermal_load                  12960 non-null  float64
 6   vib_overall_rms_mm_s          12960 non-null  float64
 7   vib_peak_mm_s                 12960 non-null  float64
 8   vib_crest_factor              12960 non-null  float64
 9   vib_kurtosis                  12960 non-null  float64
 10  vib_bpfo_amplitude            12960 non-null  float64
 11  vib_bpfi_amplitude            12960 non-null  float64
 12  pressure_differential_bar     12960 non-null  float64
 13  pressure_lub

In [38]:
model_data.isnull().sum()

temp_bearing_drive_end_c        0
temp_bearing_non_drive_end_c    0
temp_motor_winding_c            0
thermal_gradient_c_per_hr       0
temp_delta_bearing_c            0
thermal_load                    0
vib_overall_rms_mm_s            0
vib_peak_mm_s                   0
vib_crest_factor                0
vib_kurtosis                    0
vib_bpfo_amplitude              0
vib_bpfi_amplitude              0
pressure_differential_bar       0
pressure_lube_oil_bar           0
flow_rate_m3_hr                 0
cavitation_index                0
current_imbalance_pct           0
power_factor                    0
active_power_kw                 0
motor_load_pct                  0
oil_viscosity_cst               0
oil_contamination_ntu           0
hours_since_lubrication         0
operating_speed_rpm             0
efficiency_pct                  0
anomaly_label                   0
dtype: int64

In [39]:
sensor[
[
"anomaly_label",
"failure_mode_active",
"failure_mode_severity"
]
].head()

,anomaly_label,failure_mode_active,failure_mode_severity
0,0,imbalance,none
1,0,imbalance,none
2,0,imbalance,none
3,0,imbalance,none
4,0,imbalance,none


In [40]:
model_data.head()

,temp_bearing_drive_end_c,temp_bearing_non_drive_end_c,temp_motor_winding_c,thermal_gradient_c_per_hr,temp_delta_bearing_c,thermal_load,vib_overall_rms_mm_s,vib_peak_mm_s,vib_crest_factor,vib_kurtosis,...,current_imbalance_pct,power_factor,active_power_kw,motor_load_pct,oil_viscosity_cst,oil_contamination_ntu,hours_since_lubrication,operating_speed_rpm,efficiency_pct,anomaly_label
0,37.61,32.88,62.43,0.082,12.23,10.41,1.3299,3.4471,2.592,2.679,...,2.703,0.8358,46.798,47.90,45.32,9.28,902.8,758.6,84.69,0
1,46.68,43.07,78.30,9.151,21.13,22.52,1.1535,2.8962,2.511,2.609,...,1.519,0.8790,46.186,47.27,56.71,3.10,932.2,745.8,84.69,0
2,52.50,46.16,78.67,5.901,26.45,26.79,0.7931,2.0955,2.642,2.749,...,4.248,0.9473,30.621,31.34,50.83,3.85,2315.1,468.7,84.69,0
3,56.98,53.18,84.07,4.195,30.14,30.05,1.0421,3.1083,2.983,3.033,...,2.257,0.8379,38.166,39.06,44.86,6.16,1795.1,586.5,84.69,0
4,59.55,56.79,87.12,2.535,31.68,31.91,1.2432,3.6768,2.958,2.771,...,4.361,0.8689,47.114,48.22,54.46,0.12,1791.6,727.6,84.69,0


In [41]:
model_data.columns.tolist()

['temp_bearing_drive_end_c',
 'temp_bearing_non_drive_end_c',
 'temp_motor_winding_c',
 'thermal_gradient_c_per_hr',
 'temp_delta_bearing_c',
 'thermal_load',
 'vib_overall_rms_mm_s',
 'vib_peak_mm_s',
 'vib_crest_factor',
 'vib_kurtosis',
 'vib_bpfo_amplitude',
 'vib_bpfi_amplitude',
 'pressure_differential_bar',
 'pressure_lube_oil_bar',
 'flow_rate_m3_hr',
 'cavitation_index',
 'current_imbalance_pct',
 'power_factor',
 'active_power_kw',
 'motor_load_pct',
 'oil_viscosity_cst',
 'oil_contamination_ntu',
 'hours_since_lubrication',
 'operating_speed_rpm',
 'efficiency_pct',
 'anomaly_label']

Añadimos x , y para preparar el módelo

In [42]:
target = "anomaly_label"

X = model_data.drop(
    columns=[target]
)

y = model_data[target]

Verificamos las variables predictoras

In [43]:
print("Dimensión X:", X.shape)
print("Dimensión y:", y.shape)

Dimensión X: (12960, 25)
Dimensión y: (12960,)


In [44]:
X.head()

,temp_bearing_drive_end_c,temp_bearing_non_drive_end_c,temp_motor_winding_c,thermal_gradient_c_per_hr,temp_delta_bearing_c,thermal_load,vib_overall_rms_mm_s,vib_peak_mm_s,vib_crest_factor,vib_kurtosis,...,cavitation_index,current_imbalance_pct,power_factor,active_power_kw,motor_load_pct,oil_viscosity_cst,oil_contamination_ntu,hours_since_lubrication,operating_speed_rpm,efficiency_pct
0,37.61,32.88,62.43,0.082,12.23,10.41,1.3299,3.4471,2.592,2.679,...,0.1458,2.703,0.8358,46.798,47.90,45.32,9.28,902.8,758.6,84.69
1,46.68,43.07,78.30,9.151,21.13,22.52,1.1535,2.8962,2.511,2.609,...,0.1302,1.519,0.8790,46.186,47.27,56.71,3.10,932.2,745.8,84.69
2,52.50,46.16,78.67,5.901,26.45,26.79,0.7931,2.0955,2.642,2.749,...,0.0153,4.248,0.9473,30.621,31.34,50.83,3.85,2315.1,468.7,84.69
3,56.98,53.18,84.07,4.195,30.14,30.05,1.0421,3.1083,2.983,3.033,...,0.1291,2.257,0.8379,38.166,39.06,44.86,6.16,1795.1,586.5,84.69
4,59.55,56.79,87.12,2.535,31.68,31.91,1.2432,3.6768,2.958,2.771,...,0.0657,4.361,0.8689,47.114,48.22,54.46,0.12,1791.6,727.6,84.69


In [45]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: anomaly_label, dtype: int64

In [46]:
y.value_counts()

anomaly_label
0    10568
1     2392
Name: count, dtype: int64

Normalizamos

Cálculo de porcentajes donde 0- es el porcentaje de máquinas en estado normal y 1 - porcentaje de anómalias 

In [47]:
y.value_counts(normalize=True)*100

anomaly_label
0    81.54321
1    18.45679
Name: proportion, dtype: float64

Decisión técnica para el primer modelo

Entrada:
25 variables físicas del equipo

Salida:
0 → Normal
1 → Anomalía

In [48]:
import sys
import os

sys.path.append(os.path.abspath(".."))

In [49]:
import src.data_preprocessing as dp

print(dp.__file__)
print(dir(dp))

c:\Users\Stiven\Desktop\S08_26_EQUIPO_22\src\data_preprocessing.py
['__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'pd', 'prepare_sensor_data']


In [50]:


import importlib
import src.data_preprocessing as dp

importlib.reload(dp)

print(dir(dp))

['__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__spec__', 'pd', 'prepare_sensor_data']


In [51]:
from src.data_preprocessing import prepare_sensor_data

In [52]:
print(sensor.shape)

(12960, 81)


In [53]:
print(dir())

['In', 'Out', 'X', '_', '_12', '_13', '_16', '_17', '_18', '_21', '_26', '_27', '_28', '_29', '_31', '_32', '_33', '_34', '_36', '_38', '_39', '_40', '_41', '_44', '_45', '_46', '_47', '__', '___', '__builtin__', '__builtins__', '__doc__', '__loader__', '__name__', '__package__', '__spec__', '__vsc_ipynb_file__', '_dh', '_i', '_i1', '_i10', '_i11', '_i12', '_i13', '_i14', '_i15', '_i16', '_i17', '_i18', '_i19', '_i2', '_i20', '_i21', '_i22', '_i23', '_i24', '_i25', '_i26', '_i27', '_i28', '_i29', '_i3', '_i30', '_i31', '_i32', '_i33', '_i34', '_i35', '_i36', '_i37', '_i38', '_i39', '_i4', '_i40', '_i41', '_i42', '_i43', '_i44', '_i45', '_i46', '_i47', '_i48', '_i49', '_i5', '_i50', '_i51', '_i52', '_i53', '_i6', '_i7', '_i8', '_i9', '_ih', '_ii', '_iii', '_oh', 'col', 'combinations', 'comunes', 'datasets', 'df', 'df1', 'df2', 'dp', 'exit', 'failure_path', 'failures', 'features', 'get_ipython', 'hf_hub_download', 'importlib', 'model_data', 'name', 'name1', 'name2', 'open', 'os', 'pd', '

In [55]:
print(sensor.shape)

(12960, 81)


In [56]:
import os

print(os.getcwd())

c:\Users\Stiven\Desktop\S08_26_EQUIPO_22\notebooks


In [57]:
%whos DataFrame

Variable     Type         Data/Info
-----------------------------------
X            DataFrame    Shape: (12960, 25)
df           DataFrame    Shape: (1, 7)
df1          DataFrame    Shape: (12960, 81)
df2          DataFrame    Shape: (1, 7)
failures     DataFrame    Shape: (1, 7)
model_data   DataFrame    Shape: (12960, 26)
registry     DataFrame    Shape: (12, 25)
sensor       DataFrame    Shape: (12960, 81)


In [58]:
X, y = prepare_sensor_data(sensor)

In [61]:
model_data.to_parquet(
    "../data/processed/model_data.parquet",
    index=False
)

In [62]:
import pandas as pd

model_data = pd.read_parquet(
    "../data/processed/model_data.parquet"
)

In [63]:
import pandas as pd

model_data = pd.read_parquet(
    "../data/processed/model_data.parquet"
)

model_data.head()

,temp_bearing_drive_end_c,temp_bearing_non_drive_end_c,temp_motor_winding_c,thermal_gradient_c_per_hr,temp_delta_bearing_c,thermal_load,vib_overall_rms_mm_s,vib_peak_mm_s,vib_crest_factor,vib_kurtosis,...,current_imbalance_pct,power_factor,active_power_kw,motor_load_pct,oil_viscosity_cst,oil_contamination_ntu,hours_since_lubrication,operating_speed_rpm,efficiency_pct,anomaly_label
0,37.61,32.88,62.43,0.082,12.23,10.41,1.3299,3.4471,2.592,2.679,...,2.703,0.8358,46.798,47.90,45.32,9.28,902.8,758.6,84.69,0
1,46.68,43.07,78.30,9.151,21.13,22.52,1.1535,2.8962,2.511,2.609,...,1.519,0.8790,46.186,47.27,56.71,3.10,932.2,745.8,84.69,0
2,52.50,46.16,78.67,5.901,26.45,26.79,0.7931,2.0955,2.642,2.749,...,4.248,0.9473,30.621,31.34,50.83,3.85,2315.1,468.7,84.69,0
3,56.98,53.18,84.07,4.195,30.14,30.05,1.0421,3.1083,2.983,3.033,...,2.257,0.8379,38.166,39.06,44.86,6.16,1795.1,586.5,84.69,0
4,59.55,56.79,87.12,2.535,31.68,31.91,1.2432,3.6768,2.958,2.771,...,4.361,0.8689,47.114,48.22,54.46,0.12,1791.6,727.6,84.69,0


In [64]:
model_data.shape

(12960, 26)

In [65]:
model_data["anomaly_label"].value_counts()

anomaly_label
0    10568
1     2392
Name: count, dtype: int64

In [67]:
import sklearn

print(sklearn.__version__)

1.9.1


In [68]:
from src.train_model import train_anomaly_model

In [69]:
model = train_anomaly_model(
    "../data/processed/model_data.parquet"
)

Accuracy:
0.9903549382716049

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      0.99      2114
           1       0.95      1.00      0.97       478

    accuracy                           0.99      2592
   macro avg       0.98      0.99      0.98      2592
weighted avg       0.99      0.99      0.99      2592


Confusion Matrix:
[[2089   25]
 [   0  478]]


Accuary 99% no es correcto que el modelo sea tan preciso por tanto haremos una revisión

In [70]:
model.feature_importances_

array([0.02044668, 0.01355321, 0.00123348, 0.00115302, 0.00225038,
       0.0017183 , 0.06930454, 0.13797803, 0.23037131, 0.26468686,
       0.        , 0.        , 0.00352365, 0.00213886, 0.00847151,
       0.00276075, 0.0014426 , 0.01553169, 0.01407684, 0.00311329,
       0.00141832, 0.00101689, 0.00189472, 0.00378913, 0.19812593])

In [71]:
importance = pd.DataFrame({
    "feature": X.columns,
    "importance": model.feature_importances_
})

importance.sort_values(
    by="importance",
    ascending=False
)

,feature,importance
9,vib_kurtosis,0.264687
8,vib_crest_factor,0.230371
24,efficiency_pct,0.198126
7,vib_peak_mm_s,0.137978
6,vib_overall_rms_mm_s,0.069305
0,temp_bearing_drive_end_c,0.020447
17,power_factor,0.015532
18,active_power_kw,0.014077
1,temp_bearing_non_drive_end_c,0.013553
14,flow_rate_m3_hr,0.008472


Para diferenciar un equipo normal de uno con condición anómala, estas variables tuvieron mayor peso en mi decisión

| Ranking | Variable                       | Importancia |
| ------- | ------------------------------ | ----------: |
| 1       | `vib_kurtosis`                 |      26.47% |
| 2       | `vib_crest_factor`             |      23.04% |
| 3       | `efficiency_pct`               |      19.81% |
| 4       | `vib_peak_mm_s`                |      13.80% |
| 5       | `vib_overall_rms_mm_s`         |       6.93% |
| 6       | `temp_bearing_drive_end_c`     |       2.04% |
| 7       | `power_factor`                 |       1.55% |
| 8       | `active_power_kw`              |       1.41% |
| 9       | `temp_bearing_non_drive_end_c` |       1.35% |
| 10      | `flow_rate_m3_hr`              |       0.85% |


El modelo esta siendo dominado por variables de vibración la suma del conjunto representa el 70% 

Esto tiene sentido porque las señales de vibración suelen ser muy sensibles para detectar:

desgaste mecánico
desbalanceo
problemas en rodamientos
defectos de alineación

efficiency_pct encuentra relación con :
pérdida de eficiencia
condición anómala
esto es importante para el desarrollo del modelo llm 
La condición anómala está principalmente asociada con indicadores elevados de vibración, especialmente curtosis y factor de cresta, lo que puede estar relacionado con deterioro mecánico.

In [73]:
model = train_anomaly_model(
    "../data/processed/model_data.parquet"
)

Accuracy:
0.9903549382716049

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      0.99      2114
           1       0.95      1.00      0.97       478

    accuracy                           0.99      2592
   macro avg       0.98      0.99      0.98      2592
weighted avg       0.99      0.99      0.99      2592


Confusion Matrix:
[[2089   25]
 [   0  478]]
